[AWSs3 Input] Handle unexpectedEOF errors

gizas opened the pull request:
- Enhancement


## Proposed commit message


Please explain:

- WHAT: Wraps UnexpectedEOF error in the result of readjson fucntion so the caller knows it's not a permanent failure
- WHY:  To be able to retry the processing of S3 json files in case of such errors. Until now 


## Checklist


- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.


## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

- Closes https://github.com/elastic/beats/issues/42361


## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
Pinging @elastic/obs-ds-hosted-services (Team:obs-ds-hosted-services)

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

Kavindu-Dodan commented:
> We return errS3DownloadFailed error even if unexpectedEOF returns

Since you tested with the dataset that caused the error, where you able to see why this error was caused ?

> And now I use a buf to download the file and then process it. If the downloaderror happens in the first part we return the error.

Since I couldn't find the actual root cause : is it enough only to process the first part ? Can this error caused by other sections of the stream ?

faec commented on x-pack/filebeat/input/awss3/s3_objects.go:
```diff
@@ -266,7 +266,12 @@ func (p *s3ObjectProcessor) addGzipDecoderIfNeeded(body io.Reader) (io.Reader, e
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, r); err != nil {
+		// If an error occurs during the download, handle it here
+		return fmt.Errorf("%w: %w", errS3DownloadFailed, err)
```
The error type looks ok to me, this is not in the json reader itself it's in `s3ObjectProcessor` which currently is responsible for the download. The problem is that at this point in execution the download _isn't_ complete, the `io.Reader` is still a live http pipe and that's why it's able to encounter "EOF" while parsing.

An alternative that would actually separate the network issues from the parsing is to move this `io.Copy` call and its error check to `ProcessS3Object`, right after the `p.download()` call, so we are always starting with a completed download before we do any parsing. I think it would be slightly clearer that way too, so if that would address your concerns then let's try it that way.

Kavindu-Dodan replied:
Thank you  @faec for the explanation. I am fine with the error usage with the provided background. But I have another doubt with the `bytes.Buffer` usage.

With this change, we load the full object data into memory. And this could trigger a panic with `ErrTooLarge`. So I think think it is best to detect `unexpectedEOF` with existing stream reading, where the error should be thrown from the `decoder.Decode()`

faec replied:
I'm hesitant about that: if the download is successful, then loading the whole object into memory is the intent. For very large objects there is some overhead to buffering first instead of streaming directly to the decoder, but short-lived buffers are cheap for the garbage collector to recover, and if a single event is so big that we can't hold two copies of it in memory for the length of a function call then we are going to have serious issues ingesting that data regardless (since e.g. the default memory queue holds 3200 events). Most inputs load the full event data into a memory buffer before decoding it, and I think the clarity and reliability we get from doing it here is worth the extra buffer during the initial download.

In particular, I'm worried about the case I described: if a response _does_ have deterministically invalid JSON and we classify that EOF as a retryable error, then we will keep retrying that object forever despite deterministic failure.

Kavindu-Dodan replied:
>if a single event is so big that we can't hold two copies of it in memory for the length of a function call then we are going to have serious issues ingesting that data regardless 

>  Most inputs load the full event data into a memory buffer before decoding it, and I think the clarity and reliability we get from doing it here is worth the extra buffer during the initial download

+1 on these points. 

@gizas what do you think?

faec commented:
> Since I couldn't find the actual root cause : is it enough only to process the first part ? Can this error caused by other sections of the stream ?

The EOF that this is fixing is caused specifically by a dropped connection during the download process. It's still possible in theory to get that same error later, but only if the completed download contains corrupt json, and in that case we do want to return a fatal error.

Kavindu-Dodan commented:
> > Since I couldn't find the actual root cause : is it enough only to process the first part ? Can this error caused by other sections of the stream ?
> 
> The EOF that this is fixing is caused specifically by a dropped connection during the download process. It's still possible in theory to get that same error later, but only if the completed download contains corrupt json, and in that case we do want to return a fatal error.

Got it. I thought the error came from invalid JSON. 

>  It's still possible in theory to get that same error later

Since we use a `bytes.Buffer`, we are reading the full body into memory (what I understand from the `io.copy` internals and raised this concern here https://github.com/elastic/beats/pull/42420#discussion_r1960678797 ) 

> if the completed download contains corrupt json, and in that case we do want to return a fatal error.

Agree and **+1** for this.

gizas commented:
> Since you tested with the dataset that caused the error, where you able to see why this error was caused ?

I have tested it with some files that customer sent but could not reproduce the error.
As @faec said above, the unexpectedEOF we suspect that comes from an incomplete download of file. 

That is why when/ if comes we mark it again errS3DownloadFailed and retry it

> Since we use a bytes.Buffer, we are reading the full body into memory

@Kavindu-Dodan that is a total fair concern. Indeed I am also afraid and want to do some more tests. Also I am not still sure if the S3 has a size limit that might protects us - but it is not an actual thing to trust. 

So tl;dr:
- I will need to run some load tests and will come back with numbers of memory consumption
- Dont know if we need to revert the change of the buffer and just mark all this unexpectedEOF errors as retriable. This might have the drawback that we might be too agressive and report lots of errors and also keep retrying files that might overload the input. WDYT?

gizas commented:
![Screenshot 2025-02-20 at 5 46 53 PM](https://github.com/user-attachments/assets/a6c521ef-6395-4a0a-bb8b-d70b0630e6c8)
At the moment I have only run some sporadic load tests, for eg. see above 6K records in 5secs with no errors.

I will need to make tests more structured and take some memory results

Kavindu-Dodan commented:
> I will need to make tests more structured and take some memory results

Per the details in this issue of AWS Go SDK [1], I don't think it will be easy to recreate this easily. May be we might see it with large files but that's a may be.

I think it is best to focus on the workaround where we retry if we see the error. However, we should be cautious and limit retries to prevent any unintended impact. 

[^1]: https://github.com/aws/aws-sdk-go/issues/4510#issuecomment-1416406191

gizas commented:
> I think it is best to focus on the workaround where we retry if we see the error

Ok I pushed another change, reverting back to stream reader and with retries. I think is "less disruptive"
I am running various tests but still no luck to reproduce

gizas commented:
^FYI I have no idea why integration tests fail but seem related to my change. Also locally reproducible. Will keep investigating

faec commented:
Sorry, I'm not sure I understand the advantages of reverting to stream with retries. On the one hand, adding multiple retries to `readJSON` itself seems unnecessary since ephemeral errors are already retried by the caller. But on the other hand, returning a permanent error within a single bucket pass if the retry count gets too high means that ephemeral network instability can mark a valid object as failed and skip it forever. I thought the original version already solved these problems with no particular disruption, is there some other constraint I am missing?

gizas commented:
(sorry for delay was out yesterday @faec )

Indeed although the streamer does have the logic of retries, I was thinking that when the error occurs to have another retry_attempt and create again the [decoder](https://github.com/elastic/beats/pull/42420/files#diff-f345fd6a1f5ea9523117d4ead2e5f1d13fb82eb1c65a089fd34fcdd514916a96R273)

> permanent error within a single bucket pass if the retry count gets too high means that ephemeral network instability can mark a valid object as failed and skip it forever.

Indeed we dont want to do that and at least we want some retries

>  thought the original version already solved these problems with no particular disruption, i
The original version was dealing only for the S3download errors and not the current errors of unexpectedEOF. So those are the ones we try now to handle as not permanent. Long story short is that this is what I try now but again with streamer (and reverted the copy to buffer). Let me know your thoughts please?

faec commented:
I feel like there's some miscommunication... I thought the original code I reviewed was already completely correct, and the only thing I asked for was to move the buffering a little earlier for better clarity / separation of stages.

> I dont want just to return the unexpectedEOF and mark the state as false.

This isn't what I was suggesting -- I thought your original code was right. We should download the data completely into a local buffer, and any unexpectedEOF during that process should be returned as a (retryable) download error. Then after that, we should parse the buffer, and any unexpectedEOF during that stage should be returned unchanged since it is a permanent error indicating the completed download is not valid JSON.

> Indeed we dont want to do that and at least we want some retries. This is why I added the retries here

The problem is that the approach in the code now does not have a way to distinguish between an unexpectedEOF caused by the network and one caused by invalid JSON. This problem is more severe because the new retries are localized in time, so if there is temporary network stability interrupting a large download, that can appear as repeated unexpectedEOF, and flag a valid object as permanently skipped. That's why I preferred the approach with an explicit buffer, where we are always completely certain whether the error is caused by the download or the object's contents. If it comes from the download, we _want_ to retry forever (because it is valid user data and we have no way of knowing when the network issues might be fixed), and returning it as a download error means it will be retried with backoff and eventually left for the next bucket pass, so we know we won't actually lose user data. On the other hand, if the error is happening during the decoding stage, then there is no need to retry even once, since it's a deterministic failure. That's why I preferred the previous approach of having a separate buffer for the full download, which completely avoids these issues.

If it would help, we could have a quick Zoom chat to talk through the tradeoffs, since it seems like there has been some confusion about the potential problems I'm asking to address.

gizas commented:
@fae I reverted changes and now using buffer again. I was trying to test both solutions (with copy and stream) that is why I reverted them. Let me know if now looks ok.

> does not have a way to distinguish between an unexpectedEOF caused by the network and one caused by invalid JSON

If the error was coming from an invalidJSON indeed it would still fail after 3 downloads and we could return it [here](https://github.com/elastic/beats/pull/42420/commits/2034f7cbbe37676504d6bb4399e6ace24e61af4b#diff-f345fd6a1f5ea9523117d4ead2e5f1d13fb82eb1c65a089fd34fcdd514916a96R310). 

> If it comes from the download, we want to retry forever

Ok this is the small detail that wanted to see if we all agree. Then it is fine

> That's why I preferred the approach with an explicit buffer

Ok, so I have run some big file tests here (although could not reproduce the issue with unexpectedEOF). Do you have any other concerns or tests in mind?

faec reviewed (approved):
(Talked through remaining concerns offline) new revision looks good to me, thanks!

Kavindu-Dodan commented on x-pack/filebeat/input/awss3/s3_objects.go:
```diff
@@ -266,7 +266,12 @@ func (p *s3ObjectProcessor) addGzipDecoderIfNeeded(body io.Reader) (io.Reader, e
-	dec := json.NewDecoder(r)
+	var buf bytes.Buffer
+	if _, err := io.Copy(&buf, r); err != nil {
+		// If an error occurs during the download, handle it here
```
nit - should we specifically check if the error is `unexpectedEOF` ?

gizas replied:
This is how I had it in this [commit](https://github.com/elastic/beats/pull/42420/commits/cd0653c2e58a3efb3b9dad64b1569b2498eccdef#diff-f345fd6a1f5ea9523117d4ead2e5f1d13fb82eb1c65a089fd34fcdd514916a96R273), I was checking the specific unexpectedEOF error. But after our sync with @faec we decided that is better to include all errors of this phase as retry-able, because they happen in the copy phase of the buffer.

Kavindu-Dodan replied:
hmm, personally I believe we should only retry `io.ErrUnexpectedEOF` with the **assumption** that error was caused by a network error (similar to the first commit [1]). Because `io.Copy` can return many other errors. 

I will leave the decision to others here :) 


[1] - https://github.com/elastic/beats/pull/42420/commits/6c1150c2cb4e4cc4766126efd30028d6b1d9de86#diff-f345fd6a1f5ea9523117d4ead2e5f1d13fb82eb1c65a089fd34fcdd514916a96R278

gizas replied:
I dont have a strong objection. I would say for now not to revert again if you dont mind. I mean we had done this discussion and indeed although many other errors might occur in this phase of copy all should be retry-able. 
I would leave some time to @fae for another round and vote and I will try to merge from next week

faec replied:
It's true that `io.Copy` can return other errors, but none of them indicate a fundamental problem or incompatibility with the user's data, and so all of them are ultimately ephemeral from the perspective of an S3 downloader. My feeling is that if the error happened during the process of downloading the object into a local buffer, then we know it's not caused by the contents of the object (except possibly the size, e.g. an incredibly large object that hits a memory limit, but in that case it is still specific to the _current_ running state and not a permanent failure to handle the data). And I feel that failures that happen during the download process before we start looking at the data should categorically be considered retryable, since the alternative is to permanently skip valid user data, which should always be a last resort.

Kavindu-Dodan commented on CHANGELOG.next.asciidoc:
```diff
@@ -227,6 +227,7 @@ otherwise no tag is added. {issue}42208[42208] {pull}42403[42403]
 - [Journald] Fixes handling of `journalctl` restart. A known symptom was broken multiline messages when there was a restart of journalctl while aggregating the lines. {issue}41331[41331] {pull}42595[42595]
 - Fix entityanalytics activedirectory provider full sync use before initialization bug. {pull}42682[42682]
 - In the `http_endpoint` input, fix the check for a missing HMAC HTTP header. {pull}42756[42756]
+- Handle unexpectedEOF errors in aws-s3 input. Implemented retry logic on specific error. {pull}42420[42756]
```
nit - could be better if we say 

```suggestion
- Handle unexpectedEOF error in aws-s3 input and enforce retrying using download failed error {pull}42420[42756]
```

gizas commented:
@Kavindu-Dodan I will need your review as code-owner to merge :)

Kavindu-Dodan reviewed (approved):
@gizas Approving and let's keep an eye on any potential side effects
